frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -135,26 +135,86 @@ func ListTargets(ctx context.Context, dt []byte) (*targets.List, error) {
 	return l, nil
 }
 
+type namedContextState struct {
+	State llb.State
+	Image *image.Image
+
+	// State used for initializing/re-initializing the state.
+	name         string
+	out          *mutableOutput
+	copt         dockerui.ContextOpt
+	includePaths map[string]struct{}
+}
+
+// newNamedContextState returns an uninitialized namedContextState.
+func newNamedContextState(ctx context.Context, client *dockerui.Client, name string, copt dockerui.ContextOpt) (*namedContextState, error) {
+	out := &mutableOutput{}
+	nc := &namedContextState{
+		State: llb.NewState(out),
+		name:  name,
+		out:   out,
+		copt:  copt,
+	}
+	return nc.reinit(ctx, client)
+}
+
+// reinit will reinitialize this namedContextState. If it is successful, it will return itself.
+// If the new state would be invalid, it does not modify the current context and it returns nil along with
+// an error if one was present.
+func (nc *namedContextState) reinit(ctx context.Context, client *dockerui.Client) (*namedContextState, error) {
+	// Force a copy of the context options.
+	copt := nc.copt
+	if len(nc.includePaths) > 0 {
+		if includePatterns := normalizeContextPaths(nc.includePaths); includePatterns != nil {
+			copt.LocalOpts = append(copt.LocalOpts, llb.FollowPaths(includePatterns))
+		}
+	}
+	st, img, err := client.NamedContext(ctx, nc.name, copt)
```

Comment:
It would be better to avoid multiple calls to `NamedContext`. Some branches in `NamedContext` do config resolving etc.

What if `dockerui/client.NamedContext()` returns a struct that has `State/Image` properties and `SetLocalOpts([]llb.LocalOption)` method. Current`ContextOpt.LocalOpts` seems to be unused and can be replaced with the new method.

Calling `SetLocalOpts` will change the definition of `State`, either because the internal `llb.Local` is under `mutableOutput` or inside `llb.Async()`. This is only for the `llb.Local` definition though and doesn't rerun any other branches of `NamedContext()`.

Reply:
I think that could work.

Reply:
If we do that, would it make sense to also return that same struct from the `MainContext` method as well? That way we can use the same way of updating the opts later on.

Reply:
One difference is that `MainContext` doesn't ever return an `Image` (at least atm). Otherwise, I think if it is easy and unlikely to break something, we can do it now. Otherwise, we can do the namedcontext for now and look at the case for maincontext in a follow-up.

Reply:
One thing I was thinking about was what if we just returned it as part of the output similar to `mutableOutput`? Something like:

```
type localOutput struct {
    name string
    opts []llb.LocalOpts
}
```

Then have it create the vertex in an async manner. It should then be possible to inspect the output for the mounted output and update the local options.

I think this could be coded similar to `Async` or possibly coded the way async is coded.

Reply:
I'm not sure I understand where/how that `localOutput` would be used.

Reply:
Borrowing heavily from async:

```
type LocalContext struct {
	name string
	opts []llb.LocalOption
}

func (l *LocalContext) ToInput(ctx context.Context, c *llb.Constraints) (*pb.Input, error) {
	out := l.Do(ctx, c)
	return out.ToInput(ctx, c)
}

func (l *LocalContext) Vertex(ctx context.Context, c *llb.Constraints) llb.Vertex {
	out := l.Do(ctx, c)
	return out.Vertex(ctx, c)
}

func (l *LocalContext) Do(ctx context.Context, c *llb.Constraints) llb.Output {
	st := llb.Local(l.name, l.opts...)
	return st.Output()
}

func main() {
	st1 := llb.NewState(&LocalContext{
		name: "context",
	})

	st2 := st1.Dir("/src")
        // We can modify the local options by inspecting the state output.
	if lc, ok := st2.Output().(*LocalContext); ok {
		lc.opts = append(lc.opts, llb.WithCustomName("internal"))
	}

	def, err := st2.Marshal(context.TODO())
	if err != nil {
		panic(err)
	}
	_ = llb.WriteTo(def, os.Stdout)
}
```

The intention is that the areas where we have to modify the local context, such as this one:

```go
if err == nil && len(cmd.sources) == 0 {
	for _, src := range c.SourcePaths {
		d.ctxPaths[path.Join("/", filepath.ToSlash(src))] = struct{}{}
	}
}
```

This could be modified to:

```go
if err == nil {
	if len(cmd.sources) == 0 {
		for _, src := range c.SourcePaths {
			d.ctxPaths[path.Join("/", filepath.ToSlash(src))] = struct{}{}
		}
	} else if lc, ok := cmd.sources[0].state.Output().(*LocalContext); ok {
		// see above
	}
}
```

The benefit of this is we could also end up consolidating how the local options are handled so the default context and named contexts aren't different from each other.

I'll play around with the code though. I do have concerns about this being a bit too complicated to follow. Modifying the client to return a struct may end up just being easier.

Reply:
Decently certain it's possible to do this using `Async` as long as I can modify the values the client returns.

At the moment, the only way to use async for a source is to do `llb.Scratch().Async(...)` and ignore the input. Are we potentially interested in adding an `llb.Async` top-level function? I could use it and can add it as part of this change if that's helpful.

Reply:
`llb.Scratch().Async(...)` and `llb.Async` should be equivalent. The previous state (scratch in your case) is just what gets sent to the callback as the previous state value. The callback can just ignore it.

Reply:
Yea they're equivalent. I was wondering if we wanted to formally add an `llb.Async` that is a shortcut for the above.